<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Deliverables - Rail-RNA v0.1</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">
        <link href="../css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href="..">Rail-RNA v0.1</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li >
                    <a href="..">Home</a>
                </li>
            
            
            
                <li >
                    <a href="../installation/">Installation</a>
                </li>
            
            
            
                <li >
                    <a href="../tutorial/">Tutorial</a>
                </li>
            
            
            
                <li class="active">
                    <a href="./">Deliverables</a>
                </li>
            
            
            
                <li >
                    <a href="../reference/">Reference</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li >
                    <a rel="next" href="../tutorial/">
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li >
                    <a rel="prev" href="../reference/">
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#choosing-deliverables">Choosing deliverables</a></li>
        
            <li><a href="#tsv-cross-sample-matrices">tsv: cross-sample matrices</a></li>
        
            <li><a href="#idx-isofrag-index">idx: isofrag index</a></li>
        
            <li><a href="#bw-coverage-vectors">bw: coverage vectors</a></li>
        
            <li><a href="#bam-alignments">bam: alignments</a></li>
        
            <li><a href="#bed-introns-and-indels">bed: introns and indels</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h2 id="choosing-deliverables">Choosing deliverables</h2>
<p>There are several classes of outputs. Each is described in a different section below. You can toggle which outputs are written with the <code>-d/--deliverables</code> command-line parameter for <code>go</code> and <code>align</code> job flows. By default <code>--deliverables</code> is set to <code>tsv idx bam bw bed</code>, which is all the outputs described below. But for example, you might decide that you don't need the alignment BAMs. In this case, you can invoke <code>--deliverables tsv idx bw bed</code>. This will make Rail-RNA take less time to run and will also save you space---which is especially useful on S3, where storage is rented.</p>
<h3 id="tsv-cross-sample-matrices"><code>tsv</code>: cross-sample matrices</h3>
<p>These are gzip-compressed cross-sample outputs. They appear in the <code>cross_sample_results</code> subdirectory of the output directory.</p>
<ul>
<li><code>counts.tsv.gz</code>: A labeled matrix whose (i, j)th element takes the form "x<sub>ij</sub>,y<sub>ij</sub>", where x<sub>ij</sub> is the number of primary alignments in sample i to group j, and y<sub>ij</sub> is the number of uniquely aligning reads in group j. Here, a group is a chromosome, the set of unmapped reads U, the set of mapped reads M, or the set of all reads A; and "uniquely aligning read" means that there is exactly one alignment of the read with the highest alignment score. For group U, x<sub>ij</sub> = y<sub>ij</sub>. For group M, x<sub>ij</sub> is the sum of all the x<sub>ij</sub>s in the chromosome columns. For group A, x<sub>ij</sub> is the total number of reads, while y<sub>ij</sub> is the sum of the y<sub>ij</sub>s in groups U and M.</li>
<li>
<p><code>junctions.tsv.gz</code>: a labeled matrix whose (i, j)th element is the number of reads in sample j covering intron i. Introns are in the first column. Each one takes the form</p>
<pre><code>&lt;chromosome&gt;;&lt;strand (+/-)&gt;;&lt;1-based start position (inclusive)&gt;;&lt;1-based end position (exclusive)&gt;.
</code></pre>
<p>Sample names are in the first row.
  * <code>[insertion|deletion]s.tsv.gz</code>: a labeled matrix whose (i, j)th element is the number of reads in sample j covering [insertion|deletion] i. [Insertion|deletion]s are in the first column. An insertion takes the form</p>
<pre><code>&lt;chromosome&gt;;&lt;inserted base sequence&gt;;&lt;1-based position of the last base before the insertion&gt;;&lt;1-based position of the last base before the insertion&gt;
</code></pre>
<p>A deletion takes the form</p>
<pre><code>&lt;chromosome&gt;;&lt;deleted base sequence&gt;;&lt;1-based position of the first deleted base&gt;;&lt;1-based position of the first base after the deletion&gt;
</code></pre>
</li>
<li>
<p><code>normalization.tsv.gz</code>: each row takes the form <code>&lt;sample name&gt;</code>(tab)&lt;<code>--normalize-percentile</code> normalization factor for coverage vector composed of primary alignments&gt;(tab)&lt;<code>--normalize-percentile</code> normalization factor for coverage vector composed of unique alignments&gt;. The default percentile is <code>0.75</code>, corresponding to <a href="http://www.biomedcentral.com/1471-2105/11/94">upper-quartile normalization</a>.</p>
</li>
</ul>
<h3 id="idx-isofrag-index"><code>idx</code>: isofrag index</h3>
<p>This refers to the file <code>isofrags.tar.gz</code> that appears in the <code>cross_sample_results</code> subdirectory of the output directory. In general, it should be left gzipped; it can be used to align a new sample to the transcript fragments obtained from an old Rail-RNA run. Say you've aligned 500 samples from the same tissue type with Rail, and you think the list of introns you found is pretty comprehensive. Now say you just obtained another 10 samples from the same tissue type, and you want to align them. You can skip looking for novel junctions in those 10 samples by passing Rail-RNA's <code>go</code> or <code>align</code> job flow the path to this file as an argument of the <code>--isofrag-idx</code> command-line parameter.</p>
<h3 id="bw-coverage-vectors"><code>bw</code>: coverage vectors</h3>
<p>These are coverage bigWigs, and they appear in the <code>coverage_bigwigs</code> subdirectory of the output directory. There are two for each sample. One is called <code>&lt;sample name&gt;.bw</code>, and it encodes coverage of the genome by primary alignments. The other is called <code>&lt;sample name&gt;.unique.bw</code>, and it encodes coverage of the genome by uniquely aligning reads, where a uniquely aligning read has exactly one alignment with the highest alignment score. There are also coverage bigwigs for each chromosome. <code>[mean|median].&lt;chr&gt;.bw</code> encodes the [mean|median] coverage at each base on chromosome <code>&lt;chr&gt;</code> by primary alignments, and <code>[mean|median].&lt;chr&gt;.bw</code> encodes the [mean|median] coverage at each base on chromosome <code>&lt;chr&gt;</code> by uniquely aligning reads. Leo Collado-Torres describes how to use bigWigs as input to downstream analysis tools like <a href="http://bioconductor.org/packages/release/bioc/html/derfinder.html"><code>derfinder</code></a>, <a href="http://bioconductor.org/packages/release/bioc/html/edgeR.html"><code>edgeR-robust</code></a>, and <a href="http://bioconductor.org/packages/release/bioc/html/DESeq.html"><code>DESeq2</code></a> <a href="http://lcolladotor.github.io/protocols/bigwig_DEanalysis/">here</a>. Note that a bigWig is typically an order of magnitude smaller than a BAM.</p>
<h3 id="bam-alignments"><code>bam</code>: alignments</h3>
<p>These are alignment BAMs and their corresponding indexes (BAIs) that appear in the <code>alignments</code> subdirectory of the output directory. By default, there's one BAM per sample per chromosome taking the form <code>alignments.&lt;sample name&gt;.&lt;chr&gt;.bam</code>. You can use the <code>--do-not-output-bam-by-chr</code> command-line parameter for <code>go</code> and <code>align</code> job flows to output one BAM per sample. In that case, the filename takes the form <code>alignments.&lt;sample name&gt;.bam</code>. Note that outputting BAMs by chromosome is in general faster because it increases parallelism.</p>
<h3 id="bed-introns-and-indels"><code>bed</code>: introns and indels</h3>
<p>These are BED files that mirror TopHat 2's output BEDs. They appear in the <code>introns_and_indels</code> subdirectory of the output directory. For each sample <code>&lt;sample name&gt;</code>, there are three BEDs: <code>junctions.&lt;sample name&gt;.bed</code>, <code>insertions.&lt;sample name&gt;.bed</code>, and <code>deletions.&lt;sample name&gt;.bed</code>. Quotes in the following statements are from the <a href="https://ccb.jhu.edu/software/tophat/manual.shtml">TopHat 2 manual</a>. Recall that BED always uses 0-based coordinates.
  * <code>junctions.&lt;sample name&gt;.bed</code>: "Each junction consists of two connected BED blocks, where each block is as long as the maximal overhang of any read spanning the junction. The score is the number of alignments spanning the junction."
  * <code>deletions.&lt;sample name&gt;.bed</code>: "chromLeft refers to the first genomic base of the deletion."
  * <code>insertions.&lt;sample name&gt;.bed</code>: "chromLeft refers to the last genomic base before the insertion."</p></div>
        </div>

        <footer class="col-md-12">
            <hr>
            
            <center>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</center>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script src="../js/base.js"></script>
    </body>
</html>